### ISSUE TITLE
Trusted Keys TEE not working with patch in Linux TEE driver "tee: add overflow check in tee_ioctl_shm_register()"

### ISSUE THREAD

**Issue by sahilnxp** [OP-TEE/optee_os#5624]
Hi @jenswi-linaro 

I am trying to run the Trusted Keys with TEE as the backened, But I am facing and issue in tee_shm_register().
After some debugging, I found that the [tee: add overflow check in tee_ioctl_shm_register()](https://lkml.kernel.org/stable/20220822092621.3691771-1-jens.wiklander@linaro.org/#r) is causing the issue.

access_ok()  is returning an error in this case and that why tee_shm_register() is failing, If I comment access_ok() everything works fine.

I tested this on i.MX8ulp and i.MX93.

```
root@imx8ulpevk:~# keyctl add trusted kmk-trusted "new 32" @u
[  289.535519] access_ok returned error with addr = 0xffff000008397c20
[  289.541850] trusted-key-tee optee-ta-f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c: key shm register failed
add_key: Bad address
root@imx8ulpevk:~#
```

Above is the log which I got on 8ULP board.

Can you please have a look on this and let me know if there something missing at my end.

Thanks
Sahil Malhotra

**Comment by jforissier** [comment]
Hi @sahilnxp,

It might be related to static vs. dynamic shared memory. Are `CFG_CORE_DYN_SHM` and `CFG_CORE_RESERVED_SHM` enabled or disabled? Does the optee driver print the message: "optee: dynamic shared memory is enabled"?
My guess is you have CFG_CORE_RESERVED_SHM=y and CFG_CORE_DYN_SHM=n and no message. Then my reasoning is perhaps `access_ok()` cannot not return true on a range that is carved out via a `reserved-memory` node (which is how static shared memory is reserved). If that's the case you may want to try `CFG_CORE_DYN_SHM=y`.

I tried reproducing the problem with QEMUv8. Default config, except `linux/` Git rebased onto upstream v6.0.0 (because the commit you mentioned is not in our `optee` branch). Then:
```
$ make -j10 CFG_CORE_DYN_SHM=n check CHECK_TESTS=trusted-keys
[...]
Starting QEMU... done, guest is booted.
Running: keyctl tests... 
Status: keyctl tests successful
```
It claims the keyctl tests were successful but that's wrong -- @b49020 it looks like [`check_keyctl_result()`](https://github.com/OP-TEE/build/blob/3.19.0/trusted-keys.exp#L10-L34) does not catch many error cases.
Content of ../out/bin/serial0.log:
```
[...]
Welcome to Buildroot, type root or test to login
buildroot login: root
# export LD_LIBRARY_PATH=/lib:/lib/arm-linux-gnueabihf
# keyctl add trusted kmk "new 32" @u
[    7.293607] trusted-key-tee optee-ta-f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c: key shm register failed
add_key: Invalid argument
[...]
```

With the default setup however (`CFG_CORE_DYN_SHM=y`):
```
$ make -j10  check CHECK_TESTS=trusted-keys
[...]
Starting QEMU... done, guest is booted.
Running: keyctl tests... 
Status: keyctl tests successful
```
And in the `keyctl add` looks good in ../out/bin/serial0.log:
```
[...]
Welcome to Buildroot, type root or test to login
buildroot login: root
# export LD_LIBRARY_PATH=/lib:/lib/arm-linux-gnueabihf
# keyctl add trusted kmk "new 32" @u
279764142
[...]
```

**Comment by sahilnxp** [comment]
Hi @jforissier 

Sorry for not giving the full information at once.
But Dynamic shared memory is enabled:

```
root@imx8ulpevk:~# dmesg| grep optee
[    1.516068] optee: probing for conduit method.
[    1.526656] optee: revision 3.19 (e49cd872)
[    1.534546] optee: dynamic shared memory is enabled
[    1.539472] optee: initialized driver
```

**Comment by kshitizvar** [comment]
Hi @jforissier, 

Which kernel version you are using? We have tried this on 5.15 kernel, and the address we are getting is a kernel address:-**ffff0000097acc20**. And according to comments mentioned in access_ok function, it is trying to validate a user space address . That's why it is failing in our case.

Regards,
Kshitiz

**Comment by jforissier** [comment]
> Hi @jforissier,
> 
> Which kernel version you are using?

As I said I used the default OP-TEE dev environment with QEMUv8 (Aarch64): https://optee.readthedocs.io/en/latest/building/devices/qemu.html#qemu-v8

> Default config, except `linux/` Git rebased onto upstream v6.0.0 (because the commit you mentioned is not in our `optee` branch)

..so that was https://github.com/linaro-swg/linux/tree/optee (commit 29aee39cfa57) rebased onto v6.0.0.

**Comment by b49020** [comment]
> We have tried this on 5.15 kernel

So yeah the 5.15 backport for the fix was broken as mentioned here [1]. Following change should fix the 5.15 backport:

```
diff --git a/drivers/tee/tee_core.c b/drivers/tee/tee_core.c
index 3fc426dad2df..d5c4fcd8733d 100644
--- a/drivers/tee/tee_core.c
+++ b/drivers/tee/tee_core.c
@@ -334,6 +334,9 @@ tee_ioctl_shm_register(struct tee_context *ctx,
        if (data.flags)
                return -EINVAL;
 
+       if (!access_ok((void __user *)data.addr, data.length))
+               return ERR_PTR(-EFAULT);
+
        shm = tee_shm_register(ctx, data.addr, data.length,
                               TEE_SHM_DMA_BUF | TEE_SHM_USER_MAPPED);
        if (IS_ERR(shm))
diff --git a/drivers/tee/tee_shm.c b/drivers/tee/tee_shm.c
index bd96ebb82c8e..6fb4400333fb 100644
--- a/drivers/tee/tee_shm.c
+++ b/drivers/tee/tee_shm.c
@@ -223,9 +223,6 @@ struct tee_shm *tee_shm_register(struct tee_context *ctx, unsigned long addr,
                goto err;
        }
 
-       if (!access_ok((void __user *)addr, length))
-               return ERR_PTR(-EFAULT);
-
        mutex_lock(&teedev->mutex);
        shm->id = idr_alloc(&teedev->idr, shm, 1, 0, GFP_KERNEL);
        mutex_unlock(&teedev->mutex);

```

[1] https://lkml.kernel.org/stable/CAFA6WYNwk+dT_Kb3xsUQ1u5KvX+RpLwXtom6fruBbTe9W56s8Q@mail.gmail.com/

**Comment by b49020** [comment]
@jforissier 

> It claims the keyctl tests were successful but that's wrong -- @b49020 it looks like [check_keyctl_result()](https://github.com/OP-TEE/build/blob/3.19.0/trusted-keys.exp#L10-L34) does not catch many error cases.

Ah, it looks like I haven't tested it with all the negative error scenarios. I will try to spend some time to fix it but if you already have the fix, I would be happy to review.

**Comment by jforissier** [comment]
@b49020 I will propose something, thanks.

**Comment by jforissier** [comment]
https://github.com/OP-TEE/build/pull/601

**Comment by b49020** [comment]
> Following change should fix the 5.15 backport:

Here [1] is the proper v5.15 backport fixup patch.

[1] https://lore.kernel.org/stable/20221108105301.1925751-1-sumit.garg@linaro.org/

**Comment by sahilnxp** [comment]
Thanks @b49020 for the patch.
I am closing this ticket.
